File: week-5/leaf-similar-trees/leaf-similar-trees.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root1
 * @param {TreeNode} root2
 * @return {boolean}
 */
var leafSimilar = function(root1, root2) {
    const res1 = [], res2 = [];
    recurse(root1, res1);
    recurse(root2, res2);
    
    return res1.toString() === res2.toString();
};

const recurse = (root, res) => {
    if (!root) return;
    
    if (!root.left && !root.right) res.push(root.val);
    
    recurse(root.left, res);
    recurse(root.right, res)
};

File: week-5/leaf-similar-trees/leaf-similar-trees.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def leafSimilar(self, root1: TreeNode, root2: TreeNode) -> bool:
        res1 = []
        self.recurse(root1, res1)
        res2 = []
        self.recurse(root2, res2)
        
        return '~~'.join(res1) == '~~'.join(res2)
        
    def recurse(self, root, res):
        if not root:
            return
        
        if not root.left and not root.right:
            res.append(str(root.val))
            
        self.recurse(root.left, res)
        self.recurse(root.right, res)
